### ISSUE TITLE
Cannot install : undefined symbol: _ZNSt13random_device7_M_finiEv

### ISSUE THREAD

**Issue by rmcgehee** [RcppCore/RcppArmadillo#227]
I'm unable to install RcppArmadillo on my R 3.5.0 installation complied on my Red Hat 6.5 Linux box with gcc 6.3.1. RH 6.5 comes with gcc 4.4.7, which is not adequate to compile, so I've installed RH's devtoolset-6 and compiled R and all my other packages with gcc 6.3.1. However, when installing this package I get the odd message "Cannot install : undefined symbol: _ZNSt13random_device7_M_finiEv". Not sure if this is because this package is still finding one of the old gcc libraries? However, the new gcc libraries are definitely first in LD_LIBRARY_PATH. (FYI, my C is poor, so apologies if I'm not providing the best information to diagnose). Also, I tried compiling with gcc 4.9.2 as well and got the same error.

Any ideas or suggestions welcome!
```
> echo $LD_LIBRARY_PATH
/opt/rh/devtoolset-6/root/usr/lib64:/opt/rh/devtoolset-6/root/usr/lib:/srv/local/lib:/usr/local/lib

> uname -a
Linux box 2.6.32-642.3.1.el6.x86_64 #1 SMP Tue Jul 12 18:30:56 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

>R --slave -e "install.packages('RcppArmadillo')"
Installing package into '/srv/walleye/users/rmcgehee/code/rlib'
(as 'lib' is unspecified)
trying URL 'https://cran.cnr.berkeley.edu/src/contrib/RcppArmadillo_0.8.500.0.tar.gz'
Content type 'application/x-gzip' length 1347753 bytes (1.3 MB)
==================================================
downloaded 1.3 MB

* installing *source* package 'RcppArmadillo' ...
** package 'RcppArmadillo' successfully unpacked and MD5 sums checked
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C++ preprocessor... g++ -E
checking whether we are using the GNU C++ compiler... (cached) yes
checking whether g++ accepts -g... (cached) yes
checking whether g++ version is sufficient... yes, with OpenMP as version 6.3.1
checking for macOS... checking LAPACK_LIBS... R-supplied partial LAPACK found
configure: WARNING: Some complex-valued LAPACK functions may not be available
checking for OpenMP... found and suitable
configure: creating ./config.status
config.status: creating inst/include/RcppArmadilloConfigGenerated.h
config.status: creating src/Makevars
** libs
g++ -std=gnu++11 -I"/srv/local/lib64/R/include" -DNDEBUG  -I"/srv/code/rlib/Rcpp/include" -I/usr/local/include  -I../inst/include -fopenmp -fpic  -g -O2 -c RcppArmadillo.cpp -o RcppArmadillo.o
g++ -std=gnu++11 -I"/srv/local/lib64/R/include" -DNDEBUG  -I"/srv/code/rlib/Rcpp/include" -I/usr/local/include  -I../inst/include -fopenmp -fpic  -g -O2 -c RcppExports.cpp -o RcppExports.o
g++ -std=gnu++11 -I"/srv/local/lib64/R/include" -DNDEBUG  -I"/srv/code/rlib/Rcpp/include" -I/usr/local/include  -I../inst/include -fopenmp -fpic  -g -O2 -c fastLm.cpp -o fastLm.o
g++ -std=gnu++11 -shared -L/srv/local/lib64/R/lib -L/usr/local/lib64 -o RcppArmadillo.so RcppArmadillo.o RcppExports.o fastLm.o -fopenmp -L/srv/local/lib64/R/lib -lRlapack -L/srv/local/lib64/R/lib -lRblas -L/usr/lib/gcc/x86_64-redhat-linux/4.4.4 -lgfortran -lm -L/srv/local/lib64/R/lib -lR
installing to /srv/code/rlib/RcppArmadillo/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for 'RcppArmadillo' in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/srv/code/rlib/RcppArmadillo/libs/RcppArmadillo.so':
  /srv/code/rlib/RcppArmadillo/libs/RcppArmadillo.so: undefined symbol: _ZNSt13random_device7_M_finiEv
Error: loading failed
Execution halted
ERROR: loading failed
* removing '/srv/code/rlib/RcppArmadillo'

The downloaded source packages are in
	'/tmp/RtmpCJJ6qb/downloaded_packages'
Warning message:
In install.packages("RcppArmadillo") :
  installation of package 'RcppArmadillo' had non-zero exit status
```

**Comment by eddelbuettel** [comment]
That is your local problem.  RcppArmadillo works fine at CRAN, on the R Hub machines, and for thousands of users.   Moreover, the symbol is not something we should encounter as we use R's RNGs.

```sh
edd@rob:~$ c++filt _ZNSt13random_device7_M_finiEv
std::random_device::_M_fini()
edd@rob:~$ 
```

Seek local help for your machine, or install an ancient version of RcppArmadillo that matches your ancient default setup.

**Comment by eddelbuettel** [comment]
Your link line is wrong:

```sh
g++ -std=gnu++11 -shared -L/srv/local/lib64/R/lib -L/usr/local/lib64 -o RcppArmadillo.so \
    RcppArmadillo.o RcppExports.o fastLm.o -fopenmp -L/srv/local/lib64/R/lib -lRlapack \
    -L/srv/local/lib64/R/lib -lRblas -L/usr/lib/gcc/x86_64-redhat-linux/4.4.4 -lgfortran \
    -lm -L/srv/local/lib64/R/lib -lR
```

The 4.4.4 gives it away.  Maybe create a better `~/.R/Makevars` -- but it really is your local issue.

**Comment by rmcgehee** [comment]
Thanks. Not sure where the link to 4.4.4 came from, but that seems like a reasonable culprit.

FWIW, as suggested, I tried installing all the previous versions of RcppArmadillo to see how far I could get.  Version 0.7.960.1.1 (2017-08-20) works; version 0.7.960.1.2 (2017-08-27) fails, with the same error). From the NEWS file, seems that the only significant difference between the two versions is this:

. The package is now compiling under the C++11 standard (#170).

To test, I set CXX11STD= -std=c++98 in .R/Makevars and I was able to compile the most recent version without any problem. Odd.

**Comment by eddelbuettel** [comment]
Mind you, it will then branch in the code and get you C++98. You generally don't want that. 

I have this, maybe it'll help you.

```sh
edd@rob:~$ grep LD .R/Makevars 
SHLIB_CXXLD=$(CCACHE) g++$(VER) 
SHLIB_CXXLDFLAGS = $(STRIP) -shared
SHLIB_CXX11LDFLAGS = $(STRIP) -shared
SHLIB_CXX14LDFLAGS = $(STRIP) -shared
SHLIB_FCLDFLAGS = $(STRIP) -shared
SHLIB_LDFLAGS = $(STRIP) -shared
#SHLIB_CXXLD=clang
edd@rob:~$ 
```

You can ignore `$(CCACHE)` if you don't have `ccache` (you should though :) and `$(VER)` is generally empty and I can use the default compilers on Debian / Ubuntu.

**Comment by rmcgehee** [comment]
Thanks for the tip! For some reason the old gcc library path was getting added to PKG_LIBS, so setting "PKG_LIBS =" in Makevars fixed everything. (I'll try out CCache)

**Comment by honghh2018** [comment]
Hi @rmcgehee ,
     I get the same error when build the R package: RcppArmadillo 
    The error lying below:

g++ -std=gnu++11 -I"/share/apps/R/R-4.1.2-2/lib64/R/include" -DNDEBUG  -I'/share/apps/R/R-4.1.2-2/lib64/R/library/Rcpp/include' -I/usr/local/include  -I../inst/include -fopenmp -fpic  -g -O2  -c fastLm.cpp -o fastLm.o
g++ -std=gnu++11 -shared -L/share/apps/R/R-4.1.2-2/lib64/R/lib -L/usr/local/lib64 -o RcppArmadillo.so RcppArmadillo.o RcppExports.o fastLm.o -fopenmp -L/share/apps/R/R-4.1.2-2/lib64/R/lib -lRlapack -L/share/apps/R/R-4.1.2-2/lib64/R/lib -lRblas -L/usr/lib/gcc/x86_64-redhat-linux/4.8.2 -lgfortran -lm -lquadmath -L/share/apps/R/R-4.1.2-2/lib64/R/lib -lR
installing to /share/apps/R/R-4.1.2-2/lib64/R/library/00LOCK-RcppArmadillo/00new/RcppArmadillo/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘RcppArmadillo’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/share/apps/R/R-4.1.2-2/lib64/R/library/00LOCK-RcppArmadillo/00new/RcppArmadillo/libs/RcppArmadillo.so':
  /share/apps/R/R-4.1.2-2/lib64/R/library/00LOCK-RcppArmadillo/00new/RcppArmadillo/libs/RcppArmadillo.so: undefined symbol: _ZNSt19basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/share/apps/R/R-4.1.2-2/lib64/R/library/RcppArmadillo’

The downloaded source packages are in
	‘/tmp/RtmpJzmk6y/downloaded_packages’
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Warning message:
In install.packages("RcppArmadillo") :
  installation of package ‘RcppArmadillo’ had non-zero exit status

The linker get error library:

I set the correct library: /opt/rh/devtoolset-9/root/usr/lib/gcc/x86_64-redhat-linux/9 into the 
~/.R/Makevars
PKG_LIBS = /opt/rh/devtoolset-9/root/usr/lib/gcc/x86_64-redhat-linux/9
and reinstall the r package again it showing me the below error:
 
g++ -std=gnu++11 -I"/share/apps/R/R-4.1.2-3/lib64/R/include" -DNDEBUG  -I'/share/apps/R/R-4.1.2-3/lib64/R/library/Rcpp/include' -I/usr/local/include  -I../inst/include -fopenmp -fpic  -g -O2  -c fastLm.cpp -o fastLm.o
g++ -std=gnu++11 -shared -L/share/apps/R/R-4.1.2-3/lib64/R/lib -L/usr/local/lib64 -o RcppArmadillo.so RcppArmadillo.o RcppExports.o fastLm.o /opt/rh/devtoolset-9/root/usr/lib/gcc/x86_64-redhat-linux/9/ -L/share/apps/R/R-4.1.2-3/lib64/R/lib -lR
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: /opt/rh/devtoolset-9/root/usr/lib/gcc/x86_64-redhat-linux/9/: file not recognized: Is a directory
collect2: error: ld returned 1 exit status
make: *** [RcppArmadillo.so] Error 1
ERROR: compilation failed for package ‘RcppArmadillo’
* removing ‘/share/apps/R/R-4.1.2-3/lib64/R/library/RcppArmadillo’

So ,how can i fix it ?  i use the scl to install the gcc 9 version.
Any advice would be appreicated.
Best
Hanhuihong

**Comment by eddelbuettel** [comment]
Please do not hijack four year old threads, Maybe g++-9 is too old. You can always try a vintage RcppArmadillo release from the [archive section of its CRAN package](https://cran.r-project.org/src/contrib/Archive/RcppArmadillo/) that is sufficiently dated to accomodate your environment.

**Comment by conradsnicta** [comment]
@honghh2018 
> i use the scl to install the gcc 9 version.

To finalise this thread.  By stating "[scl](https://www.softwarecollections.org/)" you appear to be using RHEL 7 or CentOS 7 or one of its clones.  This is an outdated operating system, which is no longer properly supported by Red Hat.

The final release of RHEL 7 was in March 2020 (more than 2 years ago).  Consequently, RHEL stopped "Maintenance
Support 1" in August 2020.  RHEL 7 is now in a zombie state (minimal "Maintenance Support 2"), intended to allow users to migrate to RHEL 8 or RHEL 9.  https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux

In other words, if you are indeed using REHL 7 or CentOS 7, you're strongly recommended to upgrade your operating system.  This is a better solution than using "solutions" such as "scl", which are more precisely described as brittle hacks.  These hacks will cause breakage that you are experiencing.

Armadillo has been tested to work with gcc 4.8.3, which is the default on RHEL 7.  However, it's unclear whether newer versions of R (eg. 4.x), installed either via scl or manually, will work within the environment created by scl.

**Comment by conradsnicta** [comment]
@honghh2018 
> i use the scl to install the gcc 9 version.

To finalise this thread.  By stating "[scl](https://www.softwarecollections.org/)" you appear to be using RHEL 7 or CentOS 7 or one of its clones.  This is an outdated operating system, which is no longer properly supported by Red Hat.

The final release of RHEL 7 was in March 2020 (more than 2 years ago).  Consequently, RHEL stopped "Maintenance
Support 1" in August 2020.  RHEL 7 is now in a zombie state (minimal "Maintenance Support 2"), intended to allow users to migrate to RHEL 8 or RHEL 9.  https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux

If you are indeed using RHEL 7 or CentOS 7, it is strongly recommended to upgrade your operating system.  This is a much better solution than using scl. The scl approach is more precisely described as a collection of brittle hacks.  These hacks will cause breakage that you are experiencing.

Armadillo has been tested to work with gcc 4.8.3, which is the default on RHEL 7.  However, it's unclear whether newer versions of R (eg. 4.x), installed either via scl or manually, will work within the environment created by scl.
